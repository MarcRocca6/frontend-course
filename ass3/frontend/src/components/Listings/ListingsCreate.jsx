
import React, { useState } from 'react';
import Form from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';
import Row from 'react-bootstrap/Row';
import Modal from 'react-bootstrap/Modal';
import Col from 'react-bootstrap/Col';

export default function ListingsCreate () {
  const token = sessionStorage.userToken;
  const [title, setTitle] = useState('');
  const [address, setAddress] = useState('');
  const [price, setPrice] = useState(0);
  const [propertyType, setPropertyType] = useState('House');
  const [beds, setBeds] = useState(1);
  const [bedrooms, setBedrooms] = useState(1);
  const [bathrooms, setBathrooms] = useState(1);
  const [thumbnail, setThumbnail] = useState('');
  const [amenities, setAmenities] = useState([]);

  const [showModal, setShowModal] = useState(false);
  const modalClose = () => setShowModal(false);
  const modalShow = () => {
    setShowModal(true);
  }
  function setBase64 (filePath, file) {
    const reader = new FileReader();
    reader.readAsDataURL(file);
    reader.onload = function () {
      setThumbnail(reader.result)
    };
    reader.onerror = function (error) {
      console.log('Error: ', error);
    };
  }
  function error (message) {
    alert(message)
  }
  function isValidInputs () {
    if (title === '') error('Please fill in Title field')
    else if (price === '') error('Please enter a price for this hosting')
    else if (beds === '') error('Your Listing must have at least 1 bed')
    else if (bathrooms === 0) error('Your Listing must have at least 1 bathroom')
    else if (address === '') error('Please fill in Address field')
    return true
  }
  function handleSubmit (event) {
    event.preventDefault();
    if (isValidInputs()) {
      fetch('http://localhost:5005/listings/new', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: 'Bearer ' + token,
        },
        body: JSON.stringify({
          title: title,
          address: address,
          price: price,
          thumbnail: thumbnail,
          metadata:
          {
            beds: beds,
            bathrooms: bathrooms,
            bedrooms: bedrooms,
            amenities: amenities,
            type: propertyType,
          },
        })
      })
        .then(response => response.json())
        .then(data => {
          modalShow()
          console.log(data)
        });
    }
  }

  return (
    <>
      <div className="Register card text-black">
        <h1 className="text-center padding-bottom-xl" >Listings Create</h1>
        <div className="card-body p-md-3">
          <Form onSubmit={handleSubmit}>
            <Form.Group className="mb-3">
              <Form.Label>Title</Form.Label>
                <Form.Control
                  autoFocus
                  type="name"
                  placeholder=""
                  value={title}
                  onChange={(e) => setTitle(e.target.value)}
                />
            </Form.Group>
            <Form.Group className="mb-3">
              <Form.Label>Property Type</Form.Label>
              <Form.Select autoFocus value={propertyType} onChange={e => setPropertyType(e.target.value)}>
                <option>House</option>
                <option>Villa</option>
                <option>Town House</option>
                <option>Pent House</option>
                <option>Tent</option>
                <option>Mansion</option>
                <option>Appartment</option>
                <option>Granny Flat</option>
              </Form.Select>
            </Form.Group>
            <Form.Group className="mb-3">
              <Form.Label>Hosting Address</Form.Label>
              <Form.Control
                autoFocus
                type="name"
                placeholder=""
                value={address}
                onChange={(e) => setAddress(e.target.value)}
              />
            </Form.Group>
            <Form.Group className="mb-3">
              <Form.Label>Price</Form.Label>
              <Form.Control
                autoFocus
                type="number"
                placeholder="350"
                value={price}
                onChange={(e) => setPrice(e.target.value)}
              />
            </Form.Group>
            <Row className="mb-3">
              <Form.Group as={Col} controlId="formGridCity">
                <Form.Label>Beds</Form.Label>
                <Form.Control
                  autoFocus
                  type="number"
                  placeholder="1"
                  value={beds}
                  onChange={(e) => setBeds(e.target.value)}
                />
              </Form.Group>

              <Form.Group as={Col} controlId="formGridState">
                <Form.Label>Bedrooms</Form.Label>
                <Form.Control
                  autoFocus
                  type="number"
                  placeholder="1"
                  value={bedrooms}
                  onChange={(e) => setBedrooms(e.target.value)}
                />
              </Form.Group>

              <Form.Group as={Col} controlId="formGridZip">
                <Form.Label>Bathrooms</Form.Label>
                <Form.Control
                  autoFocus
                  type="number"
                  placeholder="1"
                  value={bathrooms}
                  onChange={(e) => setBathrooms(e.target.value)}
                />
              </Form.Group>
            </Row>
            <Form.Group className="mb-3">
              <Form.Label>Amenities</Form.Label>
              <Form.Select multiple value={amenities} onChange={e => setAmenities([].slice.call(e.target.selectedOptions).map(item => item.value))}>
                <option value="TV">TV</option>
                <option value="Wifi">Wifi</option>
                <option value="Pool">Pool</option>
                <option value="Patio">Patio</option>
                <option value="Parking">Parking</option>
                <option value="Hot Tub">Hot Tub</option>
                <option value="Fire pit">Fire pit</option>
                <option value="BBQ grill">BBQ grill</option>
                <option value="Pool Table">Pool Table</option>
                <option value="Indoor Fireplace">Indoor Fireplace</option>
                <option value="Outdoor Dining Area">Outdoor Dining Area</option>
              </Form.Select>
            </Form.Group>
            <Form.Group className="mb-3">
              <Form.Label>Thumbnail</Form.Label>
              <Form.Control
                type="file"
                // value={thumbnail}
                onChange={(e) => setBase64(e.target.value, e.target.files[0])}
              />
            </Form.Group>
            <Button variant="primary" type="submit">
              Create New Listing
            </Button>
            {/* <Button variant="secondary" onClick={modalShow}>
              Test out Submission Modal
            </Button> */}
          </Form>
        </div>
      </div>
      <Modal show={showModal} onHide={modalClose}>
        <Modal.Header closeButton>
          <Modal.Title>Congratulations! ðŸŽ‰ðŸ¥³</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          You have succestfully created a new listing.
        </Modal.Body>
        <Modal.Footer>
          <Button variant="secondary" onClick={modalClose}>
            Close
          </Button>
        </Modal.Footer>
      </Modal>
      <br></br>
    </>
  );
}
