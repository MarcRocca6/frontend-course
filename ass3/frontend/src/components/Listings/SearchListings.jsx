import React from 'react';
import Form from 'react-bootstrap/Form';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Container from 'react-bootstrap/Container';

import Stack from '@mui/material/Stack';
import TextField from '@mui/material/TextField';
import AdapterDateFns from '@mui/lab/AdapterDateFns';
import LocalizationProvider from '@mui/lab/LocalizationProvider';
import DesktopDatePicker from '@mui/lab/DesktopDatePicker';
import { styled } from '@mui/material/styles';
import Typography from '@mui/material/Typography';
import Slider from '@mui/material/Slider';
import Box from '@mui/material/Box';
import PropTypes from 'prop-types';

SearchListings.propTypes = {
  getState: PropTypes.array,
  setStates: PropTypes.array,
  buttonFlag: PropTypes.array
};

export default function SearchListings (props) {
  // Helper Functions
  const Separator = styled('div')(
    ({ theme }) => `height: ${theme.spacing(3)};`,
  );
  function calculateValue (value) {
    return Math.round(2 ** (value / 10));
  }

  // Change Handles
  const handleChangePriceRange = (event, newValue) => {
    props.setStates[2](newValue);
  };
  const handleChangeCheckOut = (newValue) => {
    props.setStates[1](newValue);
  };
  const handleChangeCheckIn = (newValue) => {
    props.setStates[0](newValue);
  };
  const handleToggleButton = () => {
    if (props.buttonFlag[0] === 1) props.buttonFlag[1](0);
    else if (props.buttonFlag[0] === 0) props.buttonFlag[1](1);
  };

  return (
    <div className="container">
      <Container fluid className="p-3">
        <Row>
          <Col>
            <Row>
              <Col sm={3}>
                <Form>
                  <Row>
                    <Col>
                        <TextField
                          id="outlined-basic"
                          label="Location"
                          variant="outlined"
                          onChange={(e) => props.setStates[3](e.target.value)}
                        />
                    </Col>
                    <Col sm={3} style={searchCol}>
                      <button type="button" onClick={handleToggleButton} style={searchButton} className="searchButton align-top btn btn-warning">
                        <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" style={searchIcon} className="searchIcon bi bi-search" viewBox="0 0 16 16">
                          <path d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z"/>
                        </svg>
                      </button>
                    </Col>
                  </Row>
                </Form>
              </Col>
              <Col sm={3}>
                <Form>
                  <Row>
                    <Col>
                        <TextField
                          id="outlined-basic"
                          label="Title"
                          variant="outlined"
                          onChange={(e) => props.setStates[4](e.target.value)}
                        />
                    </Col>
                    <Col sm={3} style={searchCol}>
                      <button type="button" onClick={handleToggleButton} style={searchButton} className="searchButton align-top btn btn-warning">
                        <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" style={searchIcon} className="searchIcon bi bi-search" viewBox="0 0 16 16">
                          <path d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z"/>
                        </svg>
                      </button>
                    </Col>
                  </Row>
                </Form>
              </Col>
              <Col sm={3}>
                <Form>
                  <Row>
                    <Col>
                      <LocalizationProvider dateAdapter={AdapterDateFns}>
                          <Stack spacing={3}>
                            <DesktopDatePicker
                              value={props.getState[0]}
                              label="From"
                              inputFormat="dd/MM/yyyy"
                              onChange={handleChangeCheckIn}
                              renderInput={(params) => <TextField {...params} />}
                            />
                          </Stack>
                        </LocalizationProvider>
                    </Col>
                    <Col sm={3} style={searchCol}>
                      <button type="button" onClick={handleToggleButton} style={searchButton} className="searchButton align-top btn btn-warning">
                        <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" style={searchIcon} className="searchIcon bi bi-search" viewBox="0 0 16 16">
                          <path d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z"/>
                        </svg>
                      </button>
                    </Col>
                  </Row>
                </Form>
              </Col>
              <Col sm={3}>
                <Form>
                  <Row>
                    <Col>
                      <LocalizationProvider dateAdapter={AdapterDateFns}>
                        <Stack spacing={3}>
                          <DesktopDatePicker
                            value={props.getState[1]}
                            label="To"
                            inputFormat="dd/MM/yyyy"
                            onChange={handleChangeCheckOut}
                            renderInput={(params) => <TextField {...params} />}
                          />
                        </Stack>
                      </LocalizationProvider>
                    </Col>
                    <Col sm={3} style={searchCol}>
                      <button type="button" onClick={handleToggleButton} style={searchButton} className="searchButton align-top btn btn-warning">
                        <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" style={searchIcon} className="searchIcon bi bi-search" viewBox="0 0 16 16">
                          <path d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z"/>
                        </svg>
                      </button>
                    </Col>
                  </Row>
                </Form>
              </Col>
            </Row>
          </Col>
        </Row>
        <Separator/>
        {/* <Row>
            <div id='priceHeading'>Price Range ($/night)</div>
        </Row> */}
        <Row >
          <Col sm="auto" className='align-self-center'>
            <div id='priceHeading'>Price Range</div>
            {/* <button type="button" style={searchButton} className="searchButton align-top btn btn-warning">
              <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" style={searchIcon} className="searchIcon bi bi-search" viewBox="0 0 16 16">
                <path d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z"/>
              </svg>
            </button> */}
          </Col>
          <Col>
            <Separator/>
            <Box >
              <Typography id="non-linear-slider" gutterBottom>
                {/* {calculateValue(value[0])} {calculateValue(value[1])} */}
              </Typography>
              <PrettySlider
                aria-labelledby="non-linear-slider"
                getAriaValueText={calculateValue}
                defaultValue={[50, 100]}
                marks={marks}
                scale={calculateValue}
                onChange={handleChangePriceRange}
                min={50}
                step={1}
                max={100}
                valueLabelDisplay="auto"
              />
            </Box>
          </Col>
        </Row>
      </Container>
    </div>
  );
}

const PrettySlider = styled(Slider)({
  color: '#ffc107',
  height: 8,
  '& .MuiSlider-track': {
    border: 'none',
    color: '#f4be2e',
  },
  '& .MuiSlider-thumb': {
    height: 24,
    width: 24,
    backgroundColor: '#f4be2e',
    border: '2px solid #f4be2e',
    '&:focus, &:hover, &.Mui-active, &.Mui-focusVisible': {
      boxShadow: 'inherit',
    },
    '&:hover': {
      boxShadow: '0 0 0 8px rgba(58, 133, 137, 0.16)',
    },
    '&:before': {
      display: 'none',
    },
  },
  '& .MuiSlider-valueLabel': {
    lineHeight: 1.2,
    fontSize: 12,
    background: 'unset',
    padding: 0,
    width: 32,
    height: 32,
    borderRadius: '50% 50% 50% 0',
    backgroundColor: '#757575',
    transformOrigin: 'bottom left',
    transform: 'translate(50%, -100%) rotate(-45deg) scale(0)',
    '&:before': { display: 'none' },
    '&.MuiSlider-valueLabelOpen': {
      transform: 'translate(50%, -100%) rotate(-45deg) scale(1)',
    },
    '& > *': {
      transform: 'rotate(45deg)',
    },
  },
});

const marks = [
  {
    value: 50,
    label: '$32',
  },
  {
    value: 75,
    label: '$181',
  },
  {
    value: 100,
    label: '$1024+',
  },
];

const searchCol =
  {
    paddingLeft: 0,
  }

const searchButton =
  {
    height: '100%',
  }

const searchIcon =
  {
    marginTop: -5,
  }
