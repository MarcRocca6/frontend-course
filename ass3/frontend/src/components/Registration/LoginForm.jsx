import React, { useState } from 'react';
import Form from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';
import PropTypes from 'prop-types';

LoginForm.propTypes = {
  setAuth: PropTypes.func,
  history: PropTypes.func,
};

export default function LoginForm ({ setAuth, ...props }) {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [tickbox, setTickbox] = useState('');

  /* make email case insensitive */
  // email = email.toLowerCase();

  function validateForm () {
    if (/^[a-zA-Z0-9]+@[a-zA-Z0-9]+\.[A-Za-z]+$/.test(email)) {
      if (password.length > 0) return true;
    }
    return false;
  }

  function handleSubmit (event) {
    event.preventDefault();
    fetch('http://localhost:5005/user/auth/login', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        /* make email case insensitive */
        email: email.toLowerCase(),
        password: password,
      })
    })
      .then(response => {
        if (response.ok) {
          response.json().then((data) => {
            sessionStorage.userToken = data.token;
            sessionStorage.userEmail = email;
            setAuth(data.token)
            props.history.push('/');
          });
        } else {
          response.json().then((err) => {
            console.log(err);
            alert(err.error);
          });
        }
      })
  }

  return (
    <div className="Login card text-black">
      <h1 className="text-center padding-bottom-xl" >Login</h1>
      <div className="card-body p-md-5">
        <Form onSubmit={handleSubmit}>
          <Form.Group className="mb-3" controlId="LoginEmail">
            <Form.Label>Email address</Form.Label>
            <Form.Control
              autoFocus
              type="email"
              placeholder="Enter email"
              value={email}
              onChange={(e) => setEmail(e.target.value)}
            />
            <Form.Text className="text-muted" style={{ fontSize: '0.6em' }}>
              We will never share your personal information with anyone else.
            </Form.Text>
          </Form.Group>
          <Form.Group className="mb-3" controlId="LoginPassword">
            <Form.Label>Password</Form.Label>
            <Form.Control
              type="password"
              placeholder="Password"
              value={password}
              onChange={(e) => setPassword(e.target.value)}
            />
          </Form.Group>
          <Form.Group className="mb-3" controlId="LoginCheckbox">
            <Form.Check
              type="checkbox"
              value={tickbox} onChange={(e) => setTickbox(e.target.checked)}
              label="Keep me logged in"
            />
          </Form.Group>
          <Button variant="primary" type="submit" disabled={!validateForm()}>
            Log in
          </Button>
        </Form>
      </div>
    </div>
  );
}
