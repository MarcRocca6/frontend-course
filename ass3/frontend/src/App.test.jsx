// #########################################################
// #########################################################
// ######################  Marc Tests #########3############
// #########################################################
// #########################################################

import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer'

import LoginForm from './components/Registration/LoginForm';
import SearchListings from './components/Listings/SearchListings';
import ListingsCreate from './components/Listings/ListingsCreate';

import Form from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';

describe('LoginForm', () => {
  const noop = () => {};

  it('renders without crashing when props passed in', () => {
    shallow(<LoginForm setAuth={noop} />);
  });

  it('renders correct heading', () => {
    const wrapper = shallow(<LoginForm setAuth={noop} />);
    const title = 'Login';
    const header = <h1 className="text-center padding-bottom-xl">{title}</h1>;
    expect(wrapper.contains(header)).toBe(true);
  });

  it('renders 1 button which has text "Log in"', () => {
    render(<LoginForm setAuth={noop} />);
    const items = screen.getAllByRole('button')
    expect(items).toHaveLength(1);
    expect(items[0]).toHaveTextContent('Log in');
  });

  it('button element has a succestful click feature', () => {
    const foo = jest.fn();
    const button = shallow((<Button onClick={foo}>Log in</Button>));
    button.find('button').simulate('click');
    expect(foo.mock.calls.length).toEqual(1);
  });

  it('renders 1 textbox which has placeholder text "Enter email"', () => {
    render(<LoginForm setAuth={noop} />);
    const items = screen.getAllByRole('textbox')
    expect(items).toHaveLength(1);
    expect(screen.getByPlaceholderText('Enter email') === items[0]);
  });

  it('checkbox is initally unchecked unless clicked then becomes checked', () => {
    render(<LoginForm setAuth={noop} />);
    const checkboxs = screen.getAllByRole('checkbox')
    expect(checkboxs).toHaveLength(1);
    expect(checkboxs[0].checked).toEqual(false)
    fireEvent.click(checkboxs[0])
    expect(checkboxs[0].checked).toEqual(true)
  });

  it('button initally disabled & only actives when valid email is input in form', () => {
    const { getByText } = render(<LoginForm setAuth={noop} />);
    const button = getByText(/Log in/i).closest('button');
    const email = screen.getByPlaceholderText('Enter email');
    const password = screen.getByPlaceholderText('Password');
    fireEvent.change(email, { target: { value: 'a' } })
    fireEvent.change(password, { target: { value: 'a' } })
    expect(button).toHaveAttribute('disabled');
    fireEvent.change(email, { target: { value: 'a@a.com' } })
    expect(button).not.toBeDisabled();
  });
});

describe('ViewListings', () => {
  it('renders 6 buttons, 4 are searching buttons, 2 are calendar button', () => {
    render(<SearchListings
        getState={[0, 0]}
        buttonFlag={[0, 0]}
        setStates={[0, 0, 0, 0, 0]}
      />);
    const items = screen.getAllByRole('button')
    expect(items).toHaveLength(6);
    expect(items[0].className.split(' ').includes('searchButton')).toBe(true);
    expect(items[1].className.split(' ').includes('searchButton')).toBe(true);
    expect(items[2].className.split(' ').includes('MuiButtonBase-root')).toBe(true);
    expect(items[3].className.split(' ').includes('searchButton')).toBe(true);
    expect(items[4].className.split(' ').includes('MuiButtonBase-root')).toBe(true);
    expect(items[5].className.split(' ').includes('searchButton')).toBe(true);
  });

  it('renders 2 non-linear sliders of max-val 1024 & min val 32', () => {
    render(<SearchListings
        getState={[0, 0]}
        buttonFlag={[0, 0]}
        setStates={[0, 0, 0, 0, 0]}
      />);
    const items = screen.getAllByRole('slider')
    expect(items).toHaveLength(2);
    expect(items[0]).toHaveAttribute('aria-labelledby', 'non-linear-slider');
    expect(items[1]).toHaveAttribute('aria-labelledby', 'non-linear-slider');
    expect(items[0]).toHaveAttribute('aria-valuemax', '1024');
    expect(items[1]).toHaveAttribute('aria-valuemax', '1024');
    expect(items[0]).toHaveAttribute('aria-valuemin', '32');
    expect(items[1]).toHaveAttribute('aria-valuemin', '32');
  });

  it('renders 4 textboxes, 2 for the date-inputs, 2 for strings', () => {
    render(<SearchListings
        getState={[0, 0]}
        buttonFlag={[0, 0]}
        setStates={[0, 0, 0, 0, 0]}
      />);
    const items = screen.getAllByRole('textbox')
    expect(items).toHaveLength(4);
    expect(items[0]).toHaveAttribute('id', 'outlined-basic');
    expect(items[1]).toHaveAttribute('id', 'outlined-basic');
    expect(items[2]).toHaveAttribute('placeholder', 'dd/mm/yyyy');
    expect(items[3]).toHaveAttribute('placeholder', 'dd/mm/yyyy');
  });
});

describe('CreateListing', () => {
  it('renders succestfully and the same each time', () => {
    const create = renderer.create(<ListingsCreate/>).toJSON();
    expect(create).toMatchSnapshot();
  });

  it('renders correct heading', () => {
    const wrapper = shallow(<ListingsCreate/>);
    const title = 'Listings Create';
    const header = <h1 className="text-center padding-bottom-xl">{title}</h1>;
    expect(wrapper.contains(header)).toBe(true);
  });

  it('drop down menus trigger function on change', () => {
    const _onClick = jest.fn();
    const event = { target: ['House'] };
    shallow(
      <Form.Select autoFocus onChange={e => _onClick()}>
        <option>House</option>
        <option>Villa</option>
      </Form.Select>).simulate('change', event);
    expect(_onClick).toHaveBeenCalledTimes(1)
  });

  it('number inputs trigger function on change when event is number', () => {
    const _onClick = jest.fn();
    const event = { target: [4] };
    shallow(<Form.Control autoFocus type="number"
      onChange={e => _onClick()}/>).simulate('change', event);
    expect(_onClick).toHaveBeenCalledTimes(1)
  });

  it('string inputs do not trigger function on change when event change is a string', () => {
    const _onClick = jest.fn();
    const event = { target: ['Apples'] };
    shallow(<Form.Control autoFocus type="string"
      onChange={e => _onClick()}/>).simulate('change', event);
    expect(_onClick).toHaveBeenCalledTimes(1)
  });

  it('renders 4 spinbuttons', () => {
    render(<ListingsCreate/>);
    const items = screen.getAllByRole('spinbutton')
    expect(items).toHaveLength(4);
  });
});

// #########################################################
// #########################################################
// ####################  Matthew Tests #####################
// #########################################################
// #########################################################

// import React from 'react';
// // import { render, screen } from '@testing-library/react';
// import { shallow } from 'enzyme';
// // import components to be tested
// import {NavBar} from "./components/NavBar/NavBar"

// test('renders learn react link', () => {
//   render(<App />);
//   const linkElement = screen.getByText(/learn react/i);
//   expect(linkElement).toBeInTheDocument();
// });

// describe('NavBar', () => {
//   // const noop = () => {};

//   const onClick = jest.fn();

//   it('triggers onClick event handler when clicked', () => {
//     const onClick = jest.fn();
//     shallow(<NavBar onClick={onClick} open={false} />).simulate('click');
//     expect(onClick).toBeCalledTimes(1);
//   });

// });

// following is example code from tut09

// import { shallow } from 'enzyme';
// import * as React from 'react';
// import {Menu, MenuButton, MenuItem} from './Menu';

// describe('MenuButton', () => {
//   const noop = () => {};

//   it('triggers onClick event handler when clicked', () => {
//     const onClick = jest.fn();
//     shallow(<MenuButton onClick={onClick} open={false} />).simulate('click');
//     expect(onClick).toBeCalledTimes(1);
//   });

//   it('aria-label attribute is defined', () => {
//     const button = shallow(<MenuButton onClick={noop} open={false} />);
//     expect(button.props()['aria-label']).toBeDefined();
//   });

//   it('sets aria-expanded to false when closed', () => {
//     const button = shallow(<MenuButton onClick={noop} open={false} />);
//     expect(button.props()['aria-expanded']).toBe(false);
//   });

//   it('sets aria-expanded to true when open', () => {
//     const button = shallow(<MenuButton onClick={noop} open={true} />);
//     expect(button.props()['aria-expanded']).toBe(true);
//   });
// });

// describe('MenuItem', () => {
//   const noop = () => {};

//   it('triggers onClick event handler with title when clicked', () => {
//     const onClick = jest.fn();
//     shallow(<MenuItem onClick={onClick} title={'A title'} />).simulate('click');
//     expect(onClick).toBeCalledWith('A title');
//   });

//   it('renders with custom title', () => {
//     const title = 'My custom title';
//     const button = shallow(<MenuItem onClick={noop} title={title} />);
//     expect(button.text()).toBe(title);
//   })
// });

// describe('Menu', () => {
//   const noop = () => {};
//   const items = ['Item 1', 'Item 2', 'Item 3'];

//   it('is closed by default', () => {
//     const menu = shallow(<Menu onClick={noop} items={items} />);
//     expect(menu.find(MenuButton).first()).toBeDefined();
//     expect(menu.find(MenuItem).length).toBe(0);
//   });

//   it('creates a MenuItem for every provided item', () => {
//     const menu = shallow(<Menu onClick={noop} items={items} />);
//     expect(menu.find(MenuItem).length).toBe(0);
//     menu.find(MenuButton).first().simulate('click');
//     expect(menu.find(MenuItem).length).toBe(3);
//   });
// });
