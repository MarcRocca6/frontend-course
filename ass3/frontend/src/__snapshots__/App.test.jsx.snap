// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewListings renders succestfully and the same each time 1`] = `
Array [
  <div
    className="Register card text-black"
  >
    <h1
      className="text-center padding-bottom-xl"
    >
      Listings Create
    </h1>
    <div
      className="card-body p-md-3"
    >
      <form
        className=""
        onSubmit={[Function]}
      >
        <div
          className="mb-3"
        >
          <label
            className="form-label"
          >
            Title
          </label>
          <input
            autoFocus={true}
            className="form-control"
            onChange={[Function]}
            placeholder=""
            type="name"
            value=""
          />
        </div>
        <div
          className="mb-3"
        >
          <label
            className="form-label"
          >
            Property Type
          </label>
          <select
            autoFocus={true}
            className="form-select"
            onChange={[Function]}
            value="House"
          >
            <option>
              House
            </option>
            <option>
              Villa
            </option>
            <option>
              Town House
            </option>
            <option>
              Pent House
            </option>
            <option>
              Tent
            </option>
            <option>
              Mansion
            </option>
            <option>
              Appartment
            </option>
            <option>
              Granny Flat
            </option>
          </select>
        </div>
        <div
          className="mb-3"
        >
          <label
            className="form-label"
          >
            Hosting Address
          </label>
          <input
            autoFocus={true}
            className="form-control"
            onChange={[Function]}
            placeholder=""
            type="name"
            value=""
          />
        </div>
        <div
          className="mb-3"
        >
          <label
            className="form-label"
          >
            Price
          </label>
          <input
            autoFocus={true}
            className="form-control"
            onChange={[Function]}
            placeholder="350"
            type="number"
            value={0}
          />
        </div>
        <div
          className="mb-3 row"
        >
          <div
            className="col"
          >
            <label
              className="form-label"
              htmlFor="formGridCity"
            >
              Beds
            </label>
            <input
              autoFocus={true}
              className="form-control"
              id="formGridCity"
              onChange={[Function]}
              placeholder="1"
              type="number"
              value={1}
            />
          </div>
          <div
            className="col"
          >
            <label
              className="form-label"
              htmlFor="formGridState"
            >
              Bedrooms
            </label>
            <input
              autoFocus={true}
              className="form-control"
              id="formGridState"
              onChange={[Function]}
              placeholder="1"
              type="number"
              value={1}
            />
          </div>
          <div
            className="col"
          >
            <label
              className="form-label"
              htmlFor="formGridZip"
            >
              Bathrooms
            </label>
            <input
              autoFocus={true}
              className="form-control"
              id="formGridZip"
              onChange={[Function]}
              placeholder="1"
              type="number"
              value={1}
            />
          </div>
        </div>
        <div
          className="mb-3"
        >
          <label
            className="form-label"
          >
            Amenities
          </label>
          <select
            className="form-select"
            multiple={true}
            onChange={[Function]}
            value={Array []}
          >
            <option
              value="TV"
            >
              TV
            </option>
            <option
              value="Wifi"
            >
              Wifi
            </option>
            <option
              value="Pool"
            >
              Pool
            </option>
            <option
              value="Patio"
            >
              Patio
            </option>
            <option
              value="Parking"
            >
              Parking
            </option>
            <option
              value="Hot Tub"
            >
              Hot Tub
            </option>
            <option
              value="Fire pit"
            >
              Fire pit
            </option>
            <option
              value="BBQ grill"
            >
              BBQ grill
            </option>
            <option
              value="Pool Table"
            >
              Pool Table
            </option>
            <option
              value="Indoor Fireplace"
            >
              Indoor Fireplace
            </option>
            <option
              value="Outdoor Dining Area"
            >
              Outdoor Dining Area
            </option>
          </select>
        </div>
        <div
          className="mb-3"
        >
          <label
            className="form-label"
          >
            Thumbnail
          </label>
          <input
            className="form-control"
            onChange={[Function]}
            type="file"
          />
        </div>
        <button
          className="btn btn-primary"
          disabled={false}
          type="submit"
        >
          Create New Listing
        </button>
      </form>
    </div>
  </div>,
  <br />,
]
`;
